### PR TITLE
Use CXXFLAGS; pull out default compiler settings; use COMPILE.cc

### PR DESCRIPTION
**Pull request by seantalts**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Declare copyright holder and open-source license: see below

#### Summary:
Use CXXFLAGS and implicit macros; pull out default compiler settings.
The latter should also help us use these settings more easily from
upstream Stan and CmdStan repos.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses: 
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: make/default_compiler_options [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+# - CC: The compiler to use. Expecting g++ or clang++.
+# - O: Optimization level. Valid values are {0, 1, 2, 3}.
+# - AR: archiver (must specify for cross-compiling)
+##
+CC = clang++
+O = 3
+O_STANC = 0
+AR = ar
+
+# CPPFLAGS are used for both C and C++ compilation
```
Comment (syclik) [inline make/default_compiler_options]: @seantalts, did we want to fix this properly so we're using the implicit make variables properly? If so, although using `CPPFLAGS` does work, it's actually meant to control the C preprocessor, not meant to be flags that are used for both C and C++.
https://www.gnu.org/software/make/manual/make.html#Implicit-Variables

If we're trying to be correct, I'd rather have this be `CXXFLAGS` and maybe duplicate that to `CFLAGS`.

File: make/tests [hunk 2]
```diff
@@ -15,21 +15,21 @@ $(LIBGTEST): $(LIBGTEST)(test/gtest.o)
...
-	$(COMPILE.c) -O$O $(CFLAGS_GTEST) $< $(OUTPUT_OPTION)
+	$(COMPILE.cc) -O$O $(CFLAGS_GTEST) $< $(OUTPUT_OPTION)
 
 ##
 # Rule for building a test
 ##
 test/%_test.o : test/%_test.cpp
 	@mkdir -p $(dir $@)
-	$(COMPILE.c) -O$O $(CFLAGS_GTEST) $< $(OUTPUT_OPTION)
+	$(COMPILE.cc) -O$O $(CFLAGS_GTEST) $< $(OUTPUT_OPTION)
```
Comment (syclik) [inline make/tests]: We can now get rid of this! Awesome!

We should be able to use the built-in macros if we add a target-specific variables. https://www.gnu.org/software/make/manual/html_node/Target_002dspecific.html

Did you want to try to tackle it? If you want me to try, I'll make some time later this week.

File: make/tests [hunk 3]
```diff
@@ -15,21 +15,21 @@ $(LIBGTEST): $(LIBGTEST)(test/gtest.o)
...
 
 ##
 # Rule for building a test executable
 ##
 test/%$(EXE) : test/%.o $(LIBGTEST) $(LIBCVODES)
 	@mkdir -p $(dir $@)
-	$(LINK.c) -O$O $(GTEST_MAIN) $< $(CFLAGS_GTEST) $(OUTPUT_OPTION) $(LIBGTEST) $(LDLIBS) $(LIBCVODES)
+	$(LINK.cc) -O$O $(GTEST_MAIN) $< $(CFLAGS_GTEST) $(OUTPUT_OPTION) $(LIBGTEST) $(LDLIBS) $(LIBCVODES)
```
Comment (syclik) [inline make/tests]: Same here! We can remove this target and just use the built-in macros.

File: make/tests [hunk 4]
```diff
@@ -51,10 +51,10 @@ test/%_test.d : test/%_test.cpp
 ##
 
 test/unit/multiple_translation_units%.o : test/unit/multiple_translation_units%.cpp
-	$(COMPILE.c) -O$O $< $(OUTPUT_OPTION)
+	$(COMPILE.cc) -O$O $< $(OUTPUT_OPTION)
```
Comment (syclik) [inline make/tests]: We should be able to remove this target

File: make/tests [hunk 5]
```diff
@@ -51,10 +51,10 @@ test/%_test.d : test/%_test.cpp
...
 
 test/unit/libmultiple.so : test/unit/multiple_translation_units1.o test/unit/multiple_translation_units2.o
-	$(CC) -shared -fPIC $(OUTPUT_OPTION) $^
+	$(COMPILE.cc) -shared -fPIC $(OUTPUT_OPTION) $^
```
Comment (syclik) [inline make/tests]: for this target, we should be able to just add to the linker options for this target

File: make/tests [hunk 6]
```diff
@@ -91,7 +91,7 @@ test-headers: $(HEADER_TESTS)
 ##
 test/prob/generate_tests$(EXE) : test/prob/generate_tests.cpp
 	@mkdir -p $(dir $@)
-	$(LINK.c) -O$(O_STANC) $(CFLAGS) $< $(OUTPUT_OPTION)
+	$(LINK.cc) -O$(O_STANC) $< $(OUTPUT_OPTION)
```
Comment (syclik) [inline make/tests]: remove this target

**Comment by seantalts** [comment]
This commit shows how I plan to import it upstream: https://github.com/stan-dev/stan/commit/760f33f21ab3be5dd0e2bbd321a7e77c7ecff07a

File: make/default_compiler_options (hunk 1 above)
Comment (seantalts) [inline make/default_compiler_options]: I know, but it seems to be used in this way sometimes. We couldn't duplicate CXXFLAGS to CFLAGS; the original problem is that non-C++ compilers reject CXXFLAGS like `-std=c++1y`. The alternative might be having some fourth variable that we internally use as CPPFLAGS and then copying it at the end of every makefile to CXXFLAGS and CFLAGS.
I think we should just use CPPFLAGS; I've seen this usage before in the wild. See e.g. https://stackoverflow.com/questions/2754966/cflags-vs-cppflags

File: make/tests (hunk 2 above)
Comment (seantalts) [inline make/tests]: I think switching to implicit targets could be in a new pull request and I'd be happy to let you take a swing at it as I'm not sure how to work in things like CFLAGS_GTEST to implicit targets.

**Comment by syclik** [comment]
@seantalts, before merging, can you rename the pull request to remove "implicit macros" -- this pull request doesn't actually do a lot of that and we can do that in a separate pull request.

**Comment by seantalts** [comment]
I think "implicit macros" refers to things like `COMPILE.cc` while `implicit targets` is what you're thinking of?

**Comment by seantalts** [comment]
I changed the title anyway to be more explicit. merging!

**Pull request merged**